include/tikv/Region.h

```diff
@@ -0,0 +1,200 @@
+#pragma once
+
+#include <map>
+#include <unordered_map>
+
+#include <common/Log.h>
+#include <kvproto/errorpb.pb.h>
+#include <kvproto/metapb.pb.h>
+#include <pd/Client.h>
+#include <tikv/Backoff.h>
+
+namespace pingcap
+{
+namespace kv
+{
+
+struct Store
+{
+    uint64_t id;
+    std::string addr;
+    std::map<std::string, std::string> labels;
+
+    Store() {}
+    Store(uint64_t id_, const std::string & addr_, const std::map<std::string, std::string> & labels_)
+        : id(id_), addr(addr_), labels(labels_)
+    {}
+    Store(Store &&) = default;
+    Store(const Store &) = default;
+    Store & operator=(const Store & rhs)
+    {
+        id = rhs.id;
+        addr = rhs.addr;
+        labels = rhs.labels;
+        return *this;
+    }
+};
+
+struct RegionVerID
+{
+    uint64_t id;
+    uint64_t confVer;
+    uint64_t ver;
+
+    RegionVerID(uint64_t id_, uint64_t conf_ver, uint64_t ver_) : id(id_), confVer(conf_ver), ver(ver_) {}
+
+    bool operator==(const RegionVerID & rhs) const { return id == rhs.id && confVer == rhs.confVer && ver == rhs.ver; }
+};
+
+} // namespace kv
+} // namespace pingcap
+
+namespace std
+{
+template <>
+struct hash<pingcap::kv::RegionVerID>
+{
+    using argument_type = pingcap::kv::RegionVerID;
+    using result_type = size_t;
+    size_t operator()(const pingcap::kv::RegionVerID & key) const { return key.id ^ key.confVer ^ key.ver; }
+};
+} // namespace std
+
+namespace pingcap
+{
+namespace kv
+{
+
+struct Region
+{
+    metapb::Region meta;
+    metapb::Peer peer;
+    metapb::Peer learner;
+
+    Region(const metapb::Region & meta_, const metapb::Peer & peer_) : meta(meta_), peer(peer_), learner(metapb::Peer::default_instance())
+    {}
+
+    Region(const metapb::Region & meta_, const metapb::Peer & peer_, const metapb::Peer & learner_)
+        : meta(meta_), peer(peer_), learner(learner_)
+    {}
+
+    const std::string & startKey() { return meta.start_key(); }
+
+    const std::string & endKey() { return meta.end_key(); }
+
+    bool contains(const std::string & key) { return key >= startKey() && key < endKey(); }
```

Comment:
can `endKey` be ""?

Reply:
Yes

Reply:
but `endkey==""` means `+inf`, seems it should be `key >= start && (key < end || end == "")`

Reply:
Yes, you're right.

Reply:
Fixed